File: p2.js
// ## Problem 2: Copy List with Random Pointer

// https://leetcode.com/problems/copy-list-with-random-pointer/

// time: O(n) 
// space: O(1)

const copyRandomList = function(head) {
    if (!head) return null;
    let ptr = head;
    while (ptr != null) {
      let newNode = new Node(ptr.val);
      // A->A'->B->B'->C->C'
      newNode.next = ptr.next;
      ptr.next = newNode;
      ptr = newNode.next;
    }
    ptr = head;
    // Iterate the newly created list and use the original nodes' random pointers,
    // to assign references to random pointers for cloned nodes.
    while (ptr != null) {
      ptr.next.random = (ptr.random != null) ? ptr.random.next : null;
      ptr = ptr.next.next;
    }
    // A->A'->B->B'->C->C' would be broken to A->B->C and A'->B'->C'
    let ptr_old_list = head; // A->B->C
    let ptr_new_list = head.next; // A'->B'->C'
    let head_old = head.next;
    while (ptr_old_list != null) {
      ptr_old_list.next = ptr_old_list.next.next;
      ptr_new_list.next = (ptr_new_list.next != null) ? ptr_new_list.next.next : null;
      ptr_old_list = ptr_old_list.next;
      ptr_new_list = ptr_new_list.next;
    }
    return head_old;
}

///////////////////

// Solution with Hash Map
// time: O(n) 
// space: O(n)

let visited = new Map();
function getClonedNode(node) {
    if (node != null) {
      if (visited.has(node)) return visited.get(node);
      else {
        visited.set(node, new Node(node.val, null, null));
        return visited.get(node);
      }
    }
    return null;
}
function copyRandomList(head) {
    if(head == null) return null;
    let oldNode = head;
    let newNode = new Node(oldNode.val);
    visited.set(oldNode, newNode);
    
    while (oldNode !== null) {
      newNode.random = getClonedNode(oldNode.random);
      newNode.next = getClonedNode(oldNode.next);
      oldNode = oldNode.next;
      newNode = newNode.next;
    }
    return visited.get(head);
}

File: p2.java
// Time Complexity : O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach

/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        //hashmap to keep track of node and it's deep copy
        HashMap<Node, Node> map = new HashMap<>();
        if(head == null) return null;
        Node dummy = new Node(-1);
        Node prev = dummy;
        //Traverse over all the nodes creating it's deep copy and pointing the next and random pointer
        while(head != null){
            if(!map.containsKey(head)){
                Node node = new Node(head.val);
                map.put(head,node);
            }
            if(head.random != null){
                if(!map.containsKey(head.random)){
                    map.put(head.random, new Node(head.random.val));
                }
                map.get(head).random = map.get(head.random);
            }
            prev.next = map.get(head);
            head = head.next;
            prev = prev.next;
        }

        return dummy.next;
    }
}